### PR TITLE
add J, W cache and linsolve interface for rosenbrock oop methods and add J cache for firk oop

### PR DESCRIPTION
**Pull request by huanglangwen**
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/871

### CODE REVIEW

**Comment by devmotion** [comment]
I think we should not introduce an additional implementation for generating `J` (and `W`) but rather use/modify the existing
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/cd45039e49eefdc5d60aa72b54f91e84851eb747/src/integrators/integrator_utils.jl#L386
and
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/cd45039e49eefdc5d60aa72b54f91e84851eb747/src/integrators/integrator_utils.jl#L412

**Comment by huanglangwen** [comment]
It seems the oopgenW is broken when `u` is a StaticArray. Not sure how to fix it.
https://travis-ci.org/JuliaDiffEq/OrdinaryDiffEq.jl/jobs/571800687

**Comment by huanglangwen** [comment]
It seems the oopgenW is broken when `u` is a StaticArray. Not sure how to fix it.
https://travis-ci.org/JuliaDiffEq/OrdinaryDiffEq.jl/jobs/571800687

**Comment by devmotion** [comment]
I guess it's the first time that someone actually tests the whole logic of `oop_generate_W`. I think
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/cd45039e49eefdc5d60aa72b54f91e84851eb747/src/integrators/integrator_utils.jl#L426-L432
is incorrect, since column vectors with only one entry end up in the second branch. I assume it should be changed to
```julia
J = if ndims(u) == 2 # `u` is already a matrix
        u
      else # `u` is a vector
        hcat(u, u)
      end
```
Can you check if that solves the issue?

To be honest, it still feels very crude to me, the units are incorrect and will fail if the dimension of `u` is greater than 2.

**Comment by ChrisRackauckas** [comment]
Aren't we already vecing `u` when applying the linear operator? If not, we should, and in that case `false .* rate_prototype .* rate_prototype'` should work, ever for `u` a Number.

**Comment by devmotion** [comment]
> Aren't we already vecing u when applying the linear operator? If not, we should, and in that case false .* rate_prototype .* rate_prototype' should work, ever for u a Number.

I don't see how this is related to the test failure? But I agree that probably we should just replace the snippet above and https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/cd45039e49eefdc5d60aa72b54f91e84851eb747/src/integrators/integrator_utils.jl#L438 with
```julia
J = false .* _vec(u) .* _vec(u)'
```
That would be much cleaner.

Then we could just branch depending on the type of `u` for the computation of `W` (if we do not want to compute one LU factorization, see https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/672).

**Comment by ChrisRackauckas** [comment]
The real difficulty is that we don't have an OOP linsolve interface, which we need to get in order to really complete this I think?

**Comment by devmotion** [comment]
And the fact that `lu(J)` will fail since `J` is singular, as I just noticed...

**Comment by huanglangwen** [comment]
> 
> 
> And the fact that `lu(J)` will fail since `J` is singular, as I just noticed...

Your patch works anyway. Shall I push that?

**Comment by devmotion** [comment]
> And the fact that lu(J) will fail since J is singular, as I just noticed...

Actually, it seems `lu(@SMatrix zeros(2,2))` works (and only `lu(zeros(2,2))` fails), so I think you could even write
```julia
J = false .* _vec(u) .* _vec(u)'
W = if u isa StaticArray
  lu(J)
  elseif u isa Number
    u
  else
    LU{LinearAlgebra.lutype(uEltypeNoUnits)}(Matrix{uEltypeNoUnits}(undef, 0, 0),
                                                 Vector{LinearAlgebra.BlasInt}(undef, 0),
                                                 zero(LinearAlgebra.BlasInt))
  end
```
instead of https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/1330fa7a01626a4c334d171a619802475be8d036/src/integrators/integrator_utils.jl#L432-L448

**Comment by huanglangwen** [comment]
> 
> 
> The real difficulty is that we don't have an OOP linsolve interface, which we need to get in order to really complete this I think?

What's the motivation to move from `k₃ = _reshape(W\-_vec(linsolve_tmp), axes(uprev))`?

**Comment by YingboMa** [comment]
`W\b` is wasteful to do at every stage. A better strategy is to factorize it first and reuse the factorized object. Hence we need to add an OOP linsolve interface.

**Comment by YingboMa** [comment]
> ~W\b is wasteful to do at every stage. A better strategy is to factorize it first and reuse the factorized object.~

Never mind. We want to use factorization methods other than `lu`, so we need it.

File: src/integrators/integrator_utils.jl [hunk 1]
```diff
@@ -384,59 +384,63 @@ DiffEqBase.nlsolve_f(integrator::ODEIntegrator) =
...
-      if ArrayInterface.isstructured(f.jac_prototype) || f.jac_prototype isa SparseMatrixCSC
-        J = similar(f.jac_prototype)
-        W = similar(J)
-      elseif DiffEqBase.has_jac(f) && !DiffEqBase.has_Wfact(f) && f.jac_prototype !== nothing
-        J = nothing
-        W = WOperator(f, dt, true)
-      else
-        J = false .* vec(u) .* vec(u)'
-        W = similar(J)
+  if isdefined(alg,:nlsolve)
```
Comment (YingboMa) [inline src/integrators/integrator_utils.jl]:
```suggestion
  if alg isa NewtonAlgorithm
```

File: src/caches/rosenbrock_caches.jl [hunk 2]
```diff
@@ -64,19 +64,10 @@ function alg_cache(alg::Rosenbrock23,u,rate_prototype,uEltypeNoUnits,uBottomElty
...
-    J = nothing # is J = W.J better?
-  else
-    J = false .* vec(rate_prototype) .* vec(rate_prototype)' # uEltype?
-    W = similar(J)
-  end
+  J,W = iip_generate_W(alg,u,uprev,p,t,dt,f,uEltypeNoUnits)
   tmp = zero(rate_prototype)
   atmp = similar(u, uEltypeNoUnits)
-  tab = Rosenbrock23ConstantCache(real(uBottomEltypeNoUnits),identity,identity)
+  tab = Rosenbrock23ConstantCache(real(uBottomEltypeNoUnits),identity,identity,nothing,nothing,nothing)
```
Comment (devmotion) [inline src/caches/rosenbrock_caches.jl]: I always thought that we shouldn't misuse the oop cache as tableau but rather use separate tableau structs - and even more so with the new fields for `J` and `W`.

Comment (ChrisRackauckas) [reply src/caches/rosenbrock_caches.jl]: yeah, it's really just a short hand that always worked... until it didn't.

**Comment by huanglangwen** [comment]
Is this pr good for merge? I want to start a new pr for unifying oop and iip in `derivative_utils.jl` .

File: src/caches/rosenbrock_caches.jl [hunk 3]
```diff
@@ -67,7 +67,7 @@ function alg_cache(alg::Rosenbrock23,u,rate_prototype,uEltypeNoUnits,uBottomElty
   J,W = iip_generate_W(alg,u,uprev,p,t,dt,f,uEltypeNoUnits)
   tmp = zero(rate_prototype)
   atmp = similar(u, uEltypeNoUnits)
-  tab = Rosenbrock23ConstantCache(real(uBottomEltypeNoUnits),identity,identity,nothing,nothing,nothing)
+  tab = Ros23ConstantCache(real(uBottomEltypeNoUnits))
```
Comment (devmotion) [inline src/caches/rosenbrock_caches.jl]: I think I would prefer the name `Rosenbrock23Tableau`.

Comment (huanglangwen) [reply src/caches/rosenbrock_caches.jl]: I agree, but then we need to correct all the tab struct in `\tableaus` as well as places calling the tabs in `\caches`

File: src/caches/rosenbrock_caches.jl [hunk 4]
```diff
@@ -95,7 +95,7 @@ function alg_cache(alg::Rosenbrock32,u,rate_prototype,uEltypeNoUnits,uBottomElty
   J,W = iip_generate_W(alg,u,uprev,p,t,dt,f,uEltypeNoUnits)
   tmp = zero(rate_prototype)
   atmp = similar(u, uEltypeNoUnits)
-  tab = Rosenbrock32ConstantCache(real(uBottomEltypeNoUnits),identity,identity,nothing,nothing,nothing)
+  tab = Ros32ConstantCache(real(uBottomEltypeNoUnits))
```
Comment (devmotion) [inline src/caches/rosenbrock_caches.jl]: Similar to my comment above, I think we should use something a bit more meaningful than `Ros32ConstantCache`.

File: src/caches/rosenbrock_caches.jl (hunk 3 above)
Comment (devmotion) [reply src/caches/rosenbrock_caches.jl]: Sure, there might be some other structs which probably should be renamed as well but that can be done in a different PR, I guess. For now, IMO it's sufficient to name the new struct in a reasonable way :smiley:

**Comment by ChrisRackauckas** [comment]
Let's continue the renames as another PR.  Also note that in the sphere of influence for this should be @saurabhkgp21's newest implicit extrapolation methods. For those methods, we need a safe-hatch so that way Jacobian reuse is just generally turned off. In other cases, having an easy switch on it is nice anyways, so it's something to keep in mind.

**Pull request merged**